### ISSUE TITLE
torch.nn.threshold cannot accept tensor as a threshold

### ISSUE THREAD

**Issue by isalirezag** [pytorch/pytorch#20165]
## 🚀 Feature
I think it will be useful if we can pass threshold as a tensor to the threshold function. in this way we can compute the backprop of the threshold tensor and use it in training process.

right now the threshold function just takes non-tensor threshold

**Comment by SsnL** [comment]
We could do this. But the gradient wrt to the threshold tensor will always be zero btw.

**Comment by SsnL** [comment]
Btw, isn't this just `max`?

**Comment by isalirezag** [comment]
I think that can so useful.
It wont always be zero, it might be a little complicated though...

hmmm why 'max'?

**Comment by isalirezag** [comment]
i was trying to do it with `clamp` but i couldnot make that work

**Comment by SsnL** [comment]
`max(tensor, threshold)[i]` gives you `tensor[i]` if `tensor[i] > threshold[i]`, and `threshold[i]` otherwise.

**Comment by isalirezag** [comment]
I see, yes it is kinda similar. but when you use `max` you threshold them with the max value but in the threshold function you have the ability to threshold them with whatever value you desire

**Comment by SsnL** [comment]
I think you still threshold with arbitrary values.

**Comment by isalirezag** [comment]
maybe im making a mistake. can you please provide an example?
in torch.threshold you have the freedom to choose what value to put instead of the elements that are less than the threshold.
when i tried max it was replacing the values below the threshold with the threshold value

**Comment by t-vi** [comment]
You can get the same using `torch.where(inp >= threshold, inp, value)`, if you JIT it, it'll be fast, too. To be honest, I think the `threshold != value` case (i.e. when it's not `max`) seems pretty exotic.

**Comment by isalirezag** [comment]
@t-vi `torch.where(inp >= threshold, inp, value)` will kill the gradient flow

**Comment by t-vi** [comment]
How so? You get a gradient for inp and for value, but not for threshold, but that is expected.

**Comment by isalirezag** [comment]
@t-vi 
Actually, you are right I guess.
Is it what you mean?
This will work I think:
```

A = torch.rand(1,2,3,3)
B = torch.mean(torch.rand(2,2))
C= torch.zeros(A.size())
D = torch.where(A >= B, A,C)
```

**Comment by isalirezag** [comment]
@t-vi 
Actually, you are right I guess.
Is it what you mean?

```

A = torch.rand(1,2,3,3)
B = torch.mean(torch.rand(2,2))
C= torch.zeros(A.size())
D = torch.where(A >= B, A,C)
```
